Misc improvements to STREAM sender + receiver

sappenin opened the pull request:
**Misc Updates**

* Introduce ConnectionManager into STREAM sender
* Introduce PacketLoss to `SimulatedIlpv4Network` (Fixes #279)
* Introduce Denomination class (Fixes #288)
* Enhance StreamSender to properly honor timeouts and Connection sequences per-JVM
* Enhance SenderRecieverTest to simulate various network loss scenarios.
* Add failed Link ITs to SimpleStreamSenderIT
* Formatting, Checkstyle, and Javadoc improvements

nhartner commented on stream-parent/stream-client/src/main/java/org/interledger/stream/sender/SimpleStreamSender.java:
```diff
@@ -192,70 +293,114 @@ default int totalPackets() {
+      final AtomicBoolean timeoutReached = new AtomicBoolean(false);
 
+      timeout.ifPresent($ -> {
+        ScheduledExecutorService timeoutMonitor = Executors.newSingleThreadScheduledExecutor();
```
I wonder if this causes a memory leak creating a ScheduledExecutorService but never shutting it down.

nhartner replied:
Confirmed this will actually lead to a memory leak. To avoid this, we should change the scheduled lambda from `timeoutReached.set(true)` to  `{ timeoutReached.set(true); timeoutMonitor.shutdown(); }`

nhartner commented on link-parent/link-core/src/main/java/org/interledger/link/exceptions/LinkException.java:
```diff
@@ -85,4 +85,17 @@ public LinkException(
+  public String toString() {
+    String s = getClass().getName();
+    String message = getLocalizedMessage() + " LinkId=" + getLinkId();
+    return (message != null) ? (s + ": " + message) : s;
```
I can't see now message would ever be null

sappenin replied:
Fixed here: https://github.com/hyperledger/quilt/commit/9e83dd74ad8526f0e28bea2e6a04ac880c8c5443

nhartner commented on stream-parent/stream-client/src/test/java/org/interledger/stream/sender/SendMoneyAggregatorTest.java:
```diff
@@ -52,37 +69,138 @@ public void setUp() {
     when(streamEncryptionServiceMock.decrypt(any(), any())).thenReturn(new byte[32]);
     when(linkMock.sendPacket(any())).thenReturn(mock(InterledgerRejectPacket.class));
-
+    final StreamConnection streamConnection = new StreamConnection(StreamConnectionId.of("foo"));
```
Intellij shows this as unused. Did you mean to use this or the mock?

sappenin replied:
Fixed.

nhartner commented on stream-parent/stream-client/src/main/java/org/interledger/stream/sender/StreamConnectionManager.java:
```diff
@@ -0,0 +1,75 @@
+   *
+   * @return
+   */
+  public Optional<StreamConnection> closeConnection(final StreamConnectionId streamConnectionId) {
```
Intellij says this method is never called. Seems like it should be.

sappenin replied:
See my comment above. We'll revisit this as part of #291

nhartner commented on stream-parent/stream-core/src/main/java/org/interledger/stream/StreamConnection.java:
```diff
@@ -0,0 +1,207 @@
+ * <p>Any given Connection in a JVM manages a unique sequence id that can be incremented for each Stream Packet sent
+ * over the connection.</p>
+ */
+public class StreamConnection {
```
Might consider making this implement `AutoCloseable` so that in can be used in a try-with-resources block.

sappenin replied:
Agree. Tracking in #310.

nhartner commented on stream-parent/stream-client/src/main/java/org/interledger/stream/sender/SimpleStreamSender.java:
```diff
@@ -91,95 +117,170 @@ public SimpleStreamSender(
     Objects.requireNonNull(amount);
 
-    final SendMoneyAggregator sendMoneyAggregator = new SendMoneyAggregator(
+    final StreamConnection streamConnection = this.streamConnectionManager.openConnection(
```
Per my comment above, `streamConnection` is never closed.

sappenin replied:
hmmm...I think it is closed down on [line 363](https://github.com/hyperledger/quilt/pull/305/files#diff-2ec7bd39fe6f2a1ff349260aeddbdccaR363), no?

sappenin replied:
Oh, my bad - the actual object. nm.

sappenin replied:
For `v1.0`, @nhartner and I decided that the `SimpleStreamSender` will not close the Connection nor the Stream since the only receiver is a stateless receiver. We'll revisit this as part of #291

sappenin commented:
> We should fix the minor memory leak (with ScheduledExecutor) and make sure StreamConnection#closeConnection is called in SimpleStreamSender

This is now addressed. Please re-review and approve if things look OK. 

Last night I fixed the flaky tests, but this branch is way too big, so so if there are any further substantive changes let's track them in a ticket and work on them as a fast-follower.
